video/out/hwdec/hwdec_vaapi.c

```diff
@@ -359,6 +362,43 @@ static bool try_format_map(struct ra_hwdec *hw, struct mp_image *surface)
     return ok;
 }
 
+static bool try_format_upload(struct ra_hwdec *hw, int mp_fmt)
+{
+    bool success = false;
+    if (IMGFMT_IS_HWACCEL(mp_fmt))
+        return success;
+
+    struct mp_image *src = mp_image_alloc(mp_fmt, 2, 2);
+    if (!src)
+        return success;
+
+    struct priv_owner *p = hw->priv;
+    AVBufferRef *hw_pool = av_hwframe_ctx_alloc(p->ctx->av_device_ref);
+
+    // Arbitrarily use YUV420P for the hw_subfmt. Attempt a hwupload to seee
+    // if vaExportSurfaceHandle works.
+    mp_update_av_hw_frames_pool(&hw_pool, p->ctx->av_device_ref, IMGFMT_VAAPI,
+                                IMGFMT_420P, src->w, src->h, false);
+
+    struct mp_image *dst = mp_av_pool_image_hw_upload(hw_pool, src);
+
+    VADisplay *display = p->display;
+    VADRMPRIMESurfaceDescriptor desc = {0};
+    VASurfaceID id = va_surface_id(dst);
+    VAStatus status = vaExportSurfaceHandle(display, id, VA_SURFACE_ATTRIB_MEM_TYPE_DRM_PRIME_2,
+                                            VA_EXPORT_SURFACE_COMPOSED_LAYERS | VA_EXPORT_SURFACE_READ_ONLY, &desc);
```

Comment:
Does it also fail for you if `VA_EXPORT_SURFACE_COMPOSED_LAYERS` is not set? Could it be this check? https://gitlab.freedesktop.org/mesa/mesa/-/blob/7ebb2d1dc8c244fd7e229634ce840995e2ababf9/src/gallium/frontends/va/surface.c?#L1799-1804

I'm thinking that we don't need to do whole upload process? Is it only an issue with `vaExportSurfaceHandle`? I think ffmpeg already constrain what we can upload and the upload generally does not fail.

Maybe to speed this `try` function, we could just do `vaCreateSurfaces` and `vaExportSurfaceHandle`, maybe this way would be fast enough for checking all formats?

I'm thinking that doing full upload checks, where we have APIs to query if said upload is working is bit over the top.

-----

Reply:
I'm just thinking, maybe this is ffmpeg fault that it uses memory layout that is not exportable? I mean, would it be possible to improve on this side?

As I understand, upload is working, but because memory type was selected that is not supported by `VA_EXPORT_SURFACE_COMPOSED_LAYERS` it fails to export as such. But without this flag everything works, so maybe we can patch ffmpeg to be more mindful about this possibility?

I don't know, I'm just putting idea on the table, to think about.

Reply:
I don't think it's ffmpeg's fault. It looks like to me VAAPI just lies about the formats that work.

Reply:
My point being, that nor ffmpeg or VAAPI knows that we will try to export it as `VA_EXPORT_SURFACE_COMPOSED_LAYERS ` and as I understand it works as separate planes, without this flag. So when we say that VAAPI lies, it just doesn't know how we will use it.

Reply:
They do. The gap here would be that the additional vaapi probing I wrote assumes `VA_EXPORT_SURFACE_COMPOSED_LAYERS` unconditionally and not what the VO requests.

Reply:
>assumes VA_EXPORT_SURFACE_COMPOSED_LAYERS unconditionally

I fixed that now. Do we want to make it so this extra probe is only done if the vo requests composed layers (i.e. dmabuf-wayland only)? I'm not sure if this has any relevance outside that case.

Reply:
Theoretically it can be relevant outside of that case. There are formats that can't be exported as separate layers but can be exported as composed. *sigh* Although I don't think any of these sneak through the existing sanity checks. Anyway - it seems reasonable to keep this in place for all scenarios.